Report inconsistent recording method at extension build time

**Description**

When declaring an inconsistent recorder method like this (`@Record` without recorder parameter):

```
@Record(STATIC_INIT)
public void createIndex(BuildProducer<GeneratedFileSystemResourceBuildItem> gen) throws Exception { ... }
```

Then an exception is only raised when running the extension instead of when building it:

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:1.5.1.Final:build (default) on project quarkus-lucene: Failed to build quarkus application: java.lang.RuntimeException: Failed to load steps from class dev.morling.quarkus.extensions.lucene.extension.deployment.QuarkusLuceneExtensionProcessor: public void dev.morling.quarkus.extensions.lucene.extension.deployment.QuarkusLuceneExtensionProcessor.createIndex(io.quarkus.deployment.annotations.BuildProducer) throws java.lang.Exception is marked @Record but does not inject an @Recorder object
```

--- a/core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java
+++ b/core/processor/src/main/java/io/quarkus/annotation/processor/Constants.java
@@ -49,6 +49,7 @@ final public class Constants {
 
     public static final String ANNOTATION_TEMPLATE = "io.quarkus.runtime.annotations.Template";
     public static final String ANNOTATION_RECORDER = "io.quarkus.runtime.annotations.Recorder";
+    public static final String ANNOTATION_RECORD = "io.quarkus.deployment.annotations.Record";
 
     public static final String MEMORY_SIZE_TYPE = "io.quarkus.runtime.configuration.MemorySize";
     public static final String ANNOTATION_CONFIG_ITEM = "io.quarkus.runtime.annotations.ConfigItem";

--- a/core/processor/src/main/java/io/quarkus/annotation/processor/ExtensionAnnotationProcessor.java
+++ b/core/processor/src/main/java/io/quarkus/annotation/processor/ExtensionAnnotationProcessor.java
@@ -38,6 +38,7 @@ import javax.annotation.processing.RoundEnvironment;
 import javax.lang.model.SourceVersion;
 import javax.lang.model.element.AnnotationMirror;
 import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
 import javax.lang.model.element.ExecutableElement;
 import javax.lang.model.element.Modifier;
 import javax.lang.model.element.PackageElement;
@@ -304,6 +305,7 @@ public class ExtensionAnnotationProcessor extends AbstractProcessor {
 
             final String binaryName = processingEnv.getElementUtils().getBinaryName(clazz).toString();
             if (processorClassNames.add(binaryName)) {
+                validateRecordBuildSteps(clazz);
                 recordConfigJavadoc(clazz);
                 generateAccessor(clazz);
                 final StringBuilder rbn = getRelativeBinaryName(clazz, new StringBuilder());
@@ -318,6 +320,42 @@ public class ExtensionAnnotationProcessor extends AbstractProcessor {
                     processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR,
                             "Failed to create " + rbn + " in " + pkg + ": " + e1, clazz);
                 }
+            }
+        }
+    }
+
+    private void validateRecordBuildSteps(TypeElement clazz) {
+        for (Element e : clazz.getEnclosedElements()) {
+            if (e.getKind() != ElementKind.METHOD) {
+                continue;
+            }
+            ExecutableElement ex = (ExecutableElement) e;
+            if (!isAnnotationPresent(ex, Constants.ANNOTATION_BUILD_STEP)) {
+                continue;
+            }
+            if (!isAnnotationPresent(ex, Constants.ANNOTATION_RECORD)) {
+                continue;
+            }
+
+            boolean hasRecorder = false;
+            boolean allTypesResolvable = true;
+            for (VariableElement parameter : ex.getParameters()) {
+                String parameterClassName = parameter.asType().toString();
+                TypeElement parameterTypeElement = processingEnv.getElementUtils().getTypeElement(parameterClassName);
+                if (parameterTypeElement == null) {
+                    allTypesResolvable = false;
+                } else {
+                    if (isAnnotationPresent(parameterTypeElement, Constants.ANNOTATION_RECORDER)) {
+                        hasRecorder = true;
+                        break;
+                    }
+                }
+            }
+
+            if (!hasRecorder && allTypesResolvable) {
+                processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR, "Build Step '" + clazz.getQualifiedName() + "#"
+                        + ex.getSimpleName()
+                        + "' which is annotated with '@Record' does not contain a method parameter whose type is annotated with '@Recorder'.");
             }
         }
     }

